#VRML_SIM R2020a utf8
# license: Copyright Angel Ayala. Licensed for use with research purpose onlys.
# license url: https://www.github.com/angel-ayala
# tags: static
# Transparent images sequence to simulate a fire with smoke event.

PROTO FireSmoke [
  field SFString   name             "FireSmoke"         # Is `Solid.name`.
  field SFVec3f    translation      0 0.5 0             # Is `Transform.translation`.
  field SFRotation rotation         0 0 0 0             # Is `Transform.rotation`.
  field SFFloat    fireHeight       1                   # Is `Cylinder.height`.
  field SFFloat    fireRadius       0.2                 # Is `Cylinder.radius`.
  field SFString   controller       "fire_movement"     # Is `Robot.controller`.
  field SFString   controllerArgs   ""                  # Is `Robot.controllerArgs`.
  field SFString   customData       ""                  # Is `Robot.customData`.
  field SFBool     synchronization  TRUE                # Is `Robot.synchronization`
#  field SFBool     showSmoke    TRUE
]
{
Robot {
  name IS name
  translation IS translation
  rotation IS rotation
  controller IS controller
  controllerArgs IS controllerArgs
  customData IS customData
  synchronization IS synchronization
  children [
#    %{ if fields.showSmoke.value then }%
    Display {
      translation 0 %{= fields.fireHeight.value * 0.5}% 0
      children [
        DEF SMOKE_SHAPE Shape {
          appearance PBRAppearance {
            baseColorMap ImageTexture {
              repeatS FALSE
              repeatT FALSE
            }
            roughness 1
            metalness 0
          }
          geometry Cylinder {
            bottom FALSE
            height %{= fields.fireHeight.value * 1.5}%
            radius %{= fields.fireRadius.value * 1.25}%
            top FALSE
            subdivision 8
          }
          castShadows FALSE
        }
      ]
      name "smokeDisplay"
      width 500
      height 320
    }
#    %{ end }%
    Display {
      rotation 1 0 0 0
      children [
        DEF FIRE_SHAPE Shape {
          appearance PBRAppearance {
            baseColorMap ImageTexture {
              repeatS FALSE
              repeatT FALSE
            }
            roughness 1
            metalness 0
            IBLStrength 6
          }
          geometry Cylinder {
            bottom FALSE
            height IS fireHeight
            radius IS fireRadius
            top FALSE
            subdivision 8
          }
          castShadows FALSE
        }
      ]
      name "fireDisplay"
      width 320
      height 320
    }
  ]
}
}
